clang/test/AST/HLSL/matrix-single-subscript-getter.hlsl

```diff
@@ -0,0 +1,131 @@
+// CHECK-NEXT: IntegerLiteral {{.*}} 'int' 1
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'vector<int, 4>' <LValueToRValue>
+// CHECK-NEXT: MatrixSingleSubscriptExpr {{.*}} 'vector<int, 4>' lvalue matrixcomponent
+// CHECK-NEXT: DeclRefExpr {{.*}} 'int4x4':'matrix<int, 4, 4>' lvalue ParmVar {{.*}} 'M' 'int4x4':'matrix<int, 4, 4>'
+// CHECK-NEXT: IntegerLiteral {{.*}} 'int' 2
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'vector<int, 4>' <LValueToRValue>
+// CHECK-NEXT: MatrixSingleSubscriptExpr {{.*}} 'vector<int, 4>' lvalue matrixcomponent
+// CHECK-NEXT: DeclRefExpr {{.*}} 'int4x4':'matrix<int, 4, 4>' lvalue ParmVar {{.*}} 'M' 'int4x4':'matrix<int, 4, 4>'
+// CHECK-NEXT: IntegerLiteral {{.*}} 'int' 3
+   return M[0] + M[1] + M[2] + M[3];
+}
+
+export vector<bool, 3> getBoolVecFromTemplateMat(matrix<bool, 2, 3> M) {
+    // CHECK: FunctionDecl {{.*}} used getBoolVecFromTemplateMat 'vector<bool, 3> (matrix<bool, 2, 3>)'
+    // CHECK-NEXT: ParmVarDecl {{.*}} used M 'matrix<bool, 2, 3>'
+    // CHECK-NEXT: CompoundStmt {{.*}}
+    // CHECK-NEXT: ReturnStmt {{.*}}
+    // CHECK-NEXT: ImplicitCastExpr {{.*}} 'vector<bool, 3>' <LValueToRValue>
+    // CHECK-NEXT: MatrixSingleSubscriptExpr {{.*}} 'vector<bool, 3>' lvalue matrixcomponent
+    // CHECK-NEXT: DeclRefExpr {{.*}} 'matrix<bool, 2, 3>' lvalue ParmVar {{.*}} 'M' 'matrix<bool, 2, 3>'
+    // CHECK-NEXT: IntegerLiteral {{.*}}'int' 0
+    return M[0];
+}
+
+template<typename T>
+vector<T, 3> getVecFromTemplateMat(matrix<T, 2, 3> M) {
+    // CHECK: FunctionTemplateDecl {{.*}} getVecFromTemplateMat
+    // CHECK-NEXT: TemplateTypeParmDecl {{.*}}  referenced typename depth 0 index 0 T
+    // CHECK-NEXT: FunctionDecl {{.*}} getVecFromTemplateMat 'vector<T, 3> (matrix<T, 2, 3>)'
+    // CHECK-NEXT: ParmVarDecl {{.*}} referenced M 'matrix<T, 2, 3>'
+    // CHECK-NEXT: CompoundStmt {{.*}}
+    // CHECK-NEXT: ReturnStmt {{.*}}
+    // CHECK-NEXT: MatrixSubscriptExpr {{.*}} '<incomplete matrix index type>' lvalue matrixcomponent
+    // CHECK-NEXT: DeclRefExpr {{.*}} 'matrix<T, 2, 3>' lvalue ParmVar {{.*}} 'M' 'matrix<T, 2, 3>'
+    // CHECK-NEXT: IntegerLiteral {{.*}} 'int' 0
+    // CHECK-NEXT: <<<NULL>>>
+    // CHECK-NEXT: FunctionDecl {{.*}} used getVecFromTemplateMat 'vector<bool, 3> (matrix<bool, 2, 3>)' implicit_instantiation instantiated_from {{.*}}
+    // CHECK-NEXT: TemplateArgument type 'bool'
+    // CHECK-NEXT: BuiltinType {{.*}} 'bool'
+    // CHECK-NEXT: ParmVarDecl {{.*}} used M 'matrix<bool, 2, 3>'
+    // CHECK-NEXT: CompoundStmt {{.*}}
+    // CHECK-NEXT: ReturnStmt {{.*}}
+    // CHECK-NEXT: ImplicitCastExpr {{.*}} 'vector<bool, 3>' <LValueToRValue>
+    // CHECK-NEXT: MatrixSingleSubscriptExpr {{.*}} 'vector<bool, 3>' lvalue matrixcomponent
+    // CHECK-NEXT: DeclRefExpr {{.*}} <col:12> 'matrix<bool, 2, 3>' lvalue ParmVar {{.*}} 'M' 'matrix<bool, 2, 3>'
+    // CHECK-NEXT: IntegerLiteral {{.*}} <col:14> 'int' 0
+    // CHECK-NEXT: TypedefDecl {{.*}} referenced bool3 'vector<bool, 3>'
```

Comment:
This part is the AST of the instantiated function, and it should be probably moved near the `testTemplatedMatrixAccess` section below where the instantiation happens.

Reply:
No that wouldn't  make sense The AST dump has this as a child node of `FunctionTemplateDecl`.